lib/commands/streamscreen.js

```diff
@@ -0,0 +1,405 @@
+const GST_TUTORIAL_URL = 'https://gstreamer.freedesktop.org/documentation/installing/index.html';
+const DEFAULT_HOST = '127.0.0.1';
+const TCP_HOST = '127.0.0.1';
+const DEFAULT_PORT = 8093;
+const DEFAULT_QUALITY = 70;
+const DEFAULT_BITRATE = 4000000; // Mbps
+const BOUNDARY_STRING = '--2ae9746887f170b8cf7c271047ce314c';
+
+
+async function verifyStreamingRequirements (adb) {
+  if (!_.trim(await adb.shell(['which', SCREENRECORD_BINARY]))) {
+    throw new Error(
+      `The required '${SCREENRECORD_BINARY}' binary is not available on the device under test`);
+  }
+
+  for (const binaryName of [GSTREAMER_BINARY, GST_INSPECT_BINARY]) {
+    try {
+      await fs.which(binaryName);
+    } catch (e) {
+      throw new Error(`The ${binaryName} is not available in PATH on the host system. ` +
+        `See ${GST_TUTORIAL_URL} for more details on how to install it.`);
+    }
+  }
+
+  for (const [name, modName] of _.toPairs(REQUIRED_GST_PLUGINS)) {
+    const {stdout} = await exec(GST_INSPECT_BINARY, [name]);
+    if (!_.includes(stdout, modName)) {
+      throw new Error(
+        `The required GStreamer plugin '${name}' from '${modName}' module is not installed. ` +
+        `See ${GST_TUTORIAL_URL} for more details on how to install it.`);
+    }
+  }
+}
+
+async function getDeviceInfo (adb) {
+  const output = await adb.shell(['dumpsys', 'display']);
+  const widthMatch = /\bdeviceWidth=(\d+)/.exec(output);
+  if (!widthMatch) {
+    throw new Error(`Cannot parse the device width from ${output}`);
+  }
+  const heightMatch = /\bdeviceHeight=(\d+)/.exec(output);
+  if (!heightMatch) {
+    throw new Error(`Cannot parse the device height from ${output}`);
+  }
+  const fpsMatch = /\bfps=(\d+)/.exec(output);
+  if (!fpsMatch) {
+    throw new Error(`Cannot parse the device FPS from ${output}`);
+  }
+  return {
+    width: parseInt(widthMatch[1], 10),
+    height: parseInt(heightMatch[1], 10),
+    fps: parseInt(fpsMatch[1], 10),
+  };
+}
+
+async function initDeviceStreamingProc (adb, deviceInfo, opts = {}) {
+  const {
+    width,
+    height,
+    bitRate,
+  } = opts;
+  const adjustedWidth = parseInt(width, 10) || deviceInfo.width;
+  const adjustedHeight = parseInt(height, 10) || deviceInfo.height;
+  const adjustedBitrate = parseInt(bitRate, 10) || DEFAULT_BITRATE;
+  let screenRecordCmd = SCREENRECORD_BINARY +
+    ` --output-format=h264` +
+    // 5 seconds is fine to detect rotation changes
+    ` --time-limit=${RECORDING_INTERVAL_SEC}`;
+  if (width || height) {
+    screenRecordCmd += ` --size=${adjustedWidth}x${adjustedHeight}`;
+  }
+  if (bitRate) {
+    screenRecordCmd += ` --bit-rate=${adjustedBitrate}`;
+  }
+  const adbArgs = [
+    ...adb.executable.defaultArgs,
+    'exec-out',
+    // The loop is required, because by default the maximum record duration
+    // for screenrecord is always limited
+    `while true; do ${screenRecordCmd} -; done`,
+  ];
+  const deviceStreaming = spawn(adb.executable.path, adbArgs);
```

Comment:
No way to do this command using `teen_process`?

Reply:
I did it via teen_process first, but it looks like it adds too much overhead. Here we specifically need the process stdout to be piped with minimal delay.

Reply:
Ok.